src/box/lua/access_control.lua

```diff
@@ -0,0 +1,994 @@
+local function invoke_sync()
+    -- Create a fiber channel for scheduled tasks for sync worker.
+    if not sync_tasks then
+        -- There is no good reason to have the channel capacity limited,
+        -- but tarantool fiber channels have an upper limit by design.
+        -- Thus, in rare cases when the channel is full, the mandatory
+        -- full sync is executed, dropping all other scheduled tasks.
+        sync_tasks = fiber.channel(1024)
+    else
+        -- Clear scheduled syncs on reload. Full sync will be executed
+        -- afterwards anyway.
+        while not sync_tasks:is_empty() do
+            sync_tasks:get()
+        end
+    end
+
+    if not wait_sync then
+        wait_sync = fiber.channel()
+    end
+
+    -- Set trigger on after space/function/sequence creation.
+    if not triggers_are_set then
+        triggers_are_set = true
+        box.space._user:on_replace(on_replace_trigger)
+        box.space._space:on_replace(on_replace_trigger)
+        box.space._func:on_replace(on_replace_trigger)
+        box.space._sequence:on_replace(on_replace_trigger)
+    end
+
+    -- Fiber is required here to have the credentials synced in the background
+    -- when Tarantool is in Read Only mode after it is switched to RW.
+
+    -- Note that only one fiber exists at a time.
+    if sync_credentials_fiber == nil or
+            sync_credentials_fiber:status() == 'dead' then
+        sync_credentials_fiber = fiber.new(sync_credentials_worker)
+    end
+
+    -- If Tarantool is already in Read Write mode, credentials are still
+    -- applied in the fiber to avoid possible concurrency issues, but the
+    -- main applier fiber is blocked by `wait_sync`.
+    if not box.info.ro then
+        -- Schedule a full sync with a result message on return.
+        sync_tasks:put({type = 'BLOCKING_FULL_SYNC'})
+
+        wait_sync:get()
+    else
+        -- Schedule a full sync in the background. It will be executed
+        -- when the instance switches to RW.
+        sync_tasks:put({type = 'BACKGROUND_FULL_SYNC'})
+    end
+end
+
+-- Set credentials by source and invoke full credential synchronization to set
+-- the new credentials.
+local function set(source_name, credentials)
+    if type(source_name) ~= 'string' then
+        error('Name of privilege request source must be a string', 0)
+    end
+    if credentials ~= nil and type(credentials) ~= 'table' then
+        error('Wrong format of privilege requests', 0)
+    end
+    instance_config:validate({credentials = credentials})
+    if all_creds[source_name] == nil then
+        all_creds[source_name] = {}
+        all_creds[#all_creds + 1] = all_creds[source_name]
+    end
+    all_creds[source_name].credentials = credentials or {}
+    invoke_sync()
+end
+
+-- Return credential requests by the source name. Returns all credential
+-- requests if the source name is not specified.
+local function get(source_name)
```

Comment:
For now, we don't really have a scenario for this function, maybe move it to `_internal`?